wallet/wallet-controller/src/sync/mod.rs

```diff
@@ -82,50 +102,91 @@ pub async fn sync_once<T: NodeInterface>(
+                        block_id.hex_encode()
+                    );
+
+                    Ok(())
+                },
+            )
+            .await?;
+        } else {
+            let (wallet_block_id, wallet_block_height) =
+                wallet.best_block().map_err(ControllerError::WalletError)?;
 
-        // Pause sync if connected to a node that is still downloading blocks.
-        // Otherwise the node will not be able to find a common block and the wallet will start from genesis.
-        utils::ensure!(
-            chain_info.best_block_height >= wallet_block_height,
-            ControllerError::NotEnoughBlockHeight(
+            if chain_info.best_block_id == wallet_block_id {
+                return Ok(());
+            }
+            fetch_and_sync(
+                chain_info,
+                wallet_block_id,
                 wallet_block_height,
-                chain_info.best_block_height,
+                chain_config,
+                rpc_client,
+                &mut |common_block_height: BlockHeight, block: Block| {
+                    let block_id = block.header().block_id();
+                    wallet
+                        .scan_blocks(common_block_height, vec![block])
+                        .map_err(ControllerError::WalletError)?;
+                    log::info!(
+                        "Node chainstate updated, block height: {}, tip block id: {}",
+                        common_block_height.next_height(),
+                        block_id.hex_encode()
+                    );
+
+                    Ok(())
+                },
             )
-        );
-
-        let FetchedBlock {
-            block,
-            common_block_height,
-        } = fetch_new_block(
-            chain_config,
-            rpc_client,
-            chain_info.best_block_id,
-            chain_info.best_block_height,
-            wallet_block_id,
-            wallet_block_height,
-        )
-        .await
-        .map_err(|e| ControllerError::SyncError(e.to_string()))?;
-
-        let block_id = block.header().block_id();
-        wallet
-            .scan_blocks(common_block_height, vec![block])
-            .map_err(ControllerError::WalletError)?;
-
-        log::info!(
-            "Node chainstate updated, block height: {}, tip block id: {}",
-            common_block_height.next_height(),
-            block_id.hex_encode()
-        );
+            .await?;
+        }
     }
 }
 
+async fn fetch_and_sync<T: NodeInterface>(
+    chain_info: chainstate::ChainInfo,
+    wallet_block_id: Id<GenBlock>,
+    wallet_block_height: BlockHeight,
+    chain_config: &ChainConfig,
+    rpc_client: &T,
+    wallet_sync: &mut impl FnMut(BlockHeight, Block) -> Result<(), ControllerError<T>>,
+) -> Result<(), ControllerError<T>> {
+    utils::ensure!(
+        chain_info.best_block_height >= wallet_block_height,
+        ControllerError::NotEnoughBlockHeight(wallet_block_height, chain_info.best_block_height,)
+    );
```

Comment:
I am not convinced by this check. What if there is a better chain than the current one with fewer block? The best chain rule is not necessarily based on just the number of blocks in the chain. Am I missing something?

Reply:
This check was added here to detect situations where you connect to a node that is still downloading blocks (this can happen after removing the chainstate DB). Without this, the wallet will initiate a full rescan (because the top block is not yet known), which is unnecessary in most cases. I understand that it's possible to have a better chain with a lower height, but I don't think it's a big deal. Once the same block height is reached (which should not take much time), the wallet would resume synchronization.

Reply:
No, because `is_initial_block_download` is set to false as soon as a block received that is less than 24 hours old.

Reply:
Well, can't we just use the chain-trust instead?

Reply:
We can do that in another PR, btw, but it should be done, I believe. There's no reason to depend on height when height isn't really the criterion to determine whether we need a scan.